Make fallback algorithm airtight (#236)
quicwg/load-balancers: draft-ietf-quic-load-balancers.md

## Configuration Failover {#config-failover}

In some deployments, an infrastructure will not receive traffic unless all
servers have received a configuration, and load balancers have a superset of all
configurations that are active in the server pool, thus guaranteeing that any
CID generated by a server is decodable by any load balancer. Servers and load
balancers deployed under all of these assumptions can ignore the provisions in
this subsection.

Load balancers treat connection IDs for which they have no corresponding config
ID as unroutable (see {{unroutable}}). If they have no configuration at all,
then all connection IDs are unroutable.

> Servers with no active configuration MUST issue connection IDs with the three
> most significant bits set to 0b111 to signify the connection ID is unroutable,
> corresponding to the config ID reserved for this purpose. These connection IDs

Reviewer: ```suggestion
Servers with no active configuration MUST issue connection IDs with the reserved value
of the three most significant bits set to 0b111 to signify the connection ID is unroutable.
These connection IDs
```

MUST self-encode their length (see {{length-self-description}}).

> Servers with no active configuration SHOULD NOT send the client additional

Reviewer 2: The latter. The expectation is:

1. Client sends first packets with long header: dest CID=IDCID (client chosen, random), srce CID = CID-C
2. Server respond with packets with long header: dest CID=CID-C (the SCID of client), srce CID = CID-S (server chosen)
3. Client switches to sending packets with long header: dest CID=CID-S, srce CID = CID-C
4. At some point, server sends packets with short header, dest CID = CID-C
5. At some point, client sends packets with short header, dest CID = CID-S

If the CID-S is routable, life is good. The server can provide the client with NEW CID. The system supports migration, CID rotation, NAT rebinding, maybe Multipath if negotiated.

If the CID-S is not routable, life is constrained. The load balancer picks a route for <4-tuple, d=IDCID, s=CID-C> and remembers it. When the server sees a packet from the client for <same 4-tuple, d=CID-S, s=CID-C> if associates it with the same route. When the server sees a packet from the client for <any 4-tuple, d=CID-S> it associates it with the same route again. The load balancer will not be able to associated any new destination CID with that route, because the client is not going to use long headers anymore.

ID as unroutable (see {{unroutable}}). If they have no configuration at all,
then all connection IDs are unroutable.

Servers with no active configuration MUST issue connection IDs with the reserved
value of the three most significant bits set to 0b111 to signify the connection
ID is unroutable. These connection IDs MUST self-encode their length (see
{{length-self-description}}).

Servers with no active configuration SHOULD provide the client exactly one CID

> over the life of the connection. In QUIC versions 1 and 2, therefore, servers

Reviewer: QUIC version 2 is not yet standarized. Consider omitting the `and 2`

the same 4-tuple for two connections that should be routed differently (e.g.
because they target different SNIs), as long as the packet contains a source
connection ID of nonzero length.

Therefore, the load balancer SHOULD maintain two tables that map different
values to a routing decision:

- a table indexed by a concatenation of the 4-tuple and source CID, which might
be zero-length, to route subsequent packets that do not contain the server-
generated connection ID;

- a table indexed by destination CID, if and only if it is possible for the
server to generate unroutable CIDs. This table can be shared with the one in use
for the baseline fallback algorithm.

If either table overflows, or if the load balancer loses state, it is likely the
load balancer will misroute packets.

Load balancers SHOULD maintain per-flow timers to periodically purge state in
the tables described above.

## Dropping packets

Some unroutable connection IDs are evidently client-generated if all servers
behind the load balancer are compliant with this specification. In particular,
an unroutable CID is evidently client-generated if
- it does not use the 0b111 config ID; OR

> - it uses the 0b111 config ID, it is in a long header packet, and its self-

Reviewer: Another case: if the server is encoding length, and the encoded length is known to not be a valid CID length the server would generate

perform additional checks based on that self-encoded length:

* In a long header, verify that the self-encoded length is consistent with the
CID length field in the header (i.e. the self-encoded length is one less)
* Verify that the self-encoded length is consistent with the QUIC version, if
known.
* Verify that the self-encoded length is large enough for the decoder to process
using the indicated config ID.

DCIDs that do not meet any of these criteria are routable.

### Load Balancer Forwarding {#load-balancer-forwarding}

Load balancers execute the following steps in order until one results in a
routing decision.

1. If the packet contains a routable CID, route the packet accordingly. If
there is a routing decision stored for the 4-tuple, delete the entry.
1. If the load balancer has a table of routing decisions by DCID, and the
DCID is present in it, route the packet accordingly.
1. If the load balancer has a table of routing decisions by 4-tuple, and
the DCID is present in it, route the packet accordingly.
1. Use the fallback algorithm to make a routing decision and, if applicable,
record the results in the tables indexed by 4-tuple and/or DCID.

> 

Reviewer 2: I am trying to get a picture of the underlying data model. I understand the first two steps:

1. Routable CID. That one seems obvious, but there is the issue of the Initial DCID that becomes routable "by chance". These will be assigned to the "decoded" server, regardless of the source address and source port. Is that intentional?

2. Unroutable CID that has been memorized in a global CID table. That seems obvious too. (Small question: when are such CID ejected from the table?)

3. Unroutable CID that has been memorized in a "per 4 tuple" CID table. Picoquic does something like that for the Initial DCID, because it isolate the routing from collisions due to either poor DCID (e.g., everybody pick 999...999, or everybody picks 0) or deliberate attack -- attacker sends packet with copied DCID, intercepts the connection. But doing that for other DCID than the IDCID prevents handling of NAT rebinding.

It seems weird to check the global table of ICID first, and the per tuple table second. Is that what implementations do?

Should we be explicit about the data model of the load balancer, and the tradeoff between robustness against poor random number generation of IDCID and support for NAT rebinding?

Author: > I am trying to get a picture of the underlying data model. I understand the first two steps:
> 
> 1. Routable CID. That one seems obvious, but there is the issue of the Initial DCID that becomes routable "by chance". These will be assigned to the "decoded" server, regardless of the source address and source port. Is that intentional?

Yes. This is extremely unlikely because (1) clients sending 8B CIDs is a de facto standard, and QUIC-LB will rarely come out to 8B and (2) even if so, the right thing to do is have a sparse server ID result (as described in S3.3) so that even if it decodes, the server ID is invalid.

> 2. Unroutable CID that has been memorized in a global CID table. That seems obvious too. (Small question: when are such CID ejected from the table?)

we chatted offline and agreed that it would be possible for an attacker to eject people from the table if it was on a routable CID, so ejection should be timer based.

> 3. Unroutable CID that has been memorized in a "per 4 tuple" CID table. Picoquic does something like that for the Initial DCID, because it isolate the routing from collisions due to either poor DCID (e.g., everybody pick 999...999, or everybody picks 0) or deliberate attack -- attacker sends packet with copied DCID, intercepts the connection. But doing that for other DCID than the IDCID prevents handling of NAT rebinding.
> 
> It seems weird to check the global table of ICID first, and the per tuple table second. Is that what implementations do?

There are no implementations of this PR! As you mentioned in our discussion, it's possible to have multiple connection IDs on the same connection.

> 
> Should we be explicit about the data model of the load balancer, and the tradeoff between robustness against poor random number generation of IDCID and support for NAT rebinding?

Online we discussed a major rewrite, which I'll summarize in another comment.

* The config rotation bits ({{config-rotation}}) may not correspond to an active
configuration. Note: a packet with a DCID with config ID codepoint 0b111 (see
{{config-failover}}) is always unroutable.
* If the packet header encodes the DCID length, the DCID might not be long
enough for the decoder to process.
* The extracted server mapping might not correspond to an active server.

If the load balancer has knowledge that all servers in the pool are encoding
CID length in the first octet (see {{length-self-description}}), additional
checks are possible based on that self-encoded length:

* It does not match the CID length in a long header.
* It is not valid for a QUIC version the packet is known to
encode.
* It is too short for the decoder to process using the indicated config ID.

DCIDs that do not meet any of these criteria are routable.

> 

Reviewer 2: This a bit awkward. Start with "additional checks are possible" -- the dreaded passive form. I would rather say: 

_If the load balancer has  knowledge that all servers in the pool are encoding CID
 length in the first octet (see {{length-self-description}}), it MAY perform additional
 checks based on that self-encoded length:_

Then, the following tests should be written in the positive way:

_* In a long header, verify that the self-encoded length is consistent with the CID length,_
_* Verify that the self-encoded length is consistent with the QUIC version,_
_* Verify that the self-encoded length is large enough for the decoder to process using the indicated config ID._

_DCIDs that do not meet these criteria are not routable._

I have an additional question regarding the version check. Is this something to be done only for long headers, which include the version number, or for all packets? If for all packets, how does the LB knows the negotiated version number?

Author: Done.

I extended your second bullet to say "consistent with the QUIC versions, if known." I would expect this check for long headers; for short headers, it might keep state.

that related packets go to the same server.

A fallback algorithm MAY parse a packet from a QUIC version it understands and
use that information to make a routing decision. If so, it MAY buffer packets
with unroutable DCIDs to await further packets that allow it to make a routing
decision. For example, it may use the TLS 1.3 Server Name Indication (SNI)
field, which might appear in one of several QUIC version 1 Initial packets.

> 

Reviewer 2: What about Post-Quantum? Do you mean, buffer the initial packets until the SNI can be retrieved from the Client Hello contained in the Crypto frames"? What if the SNI straddles multiple QUIC packets? What if ECH is used?

Author: I think that's what this is saying, right? "SNI field, which might appear in one of several... packets". Is there another wording you would suggest/

As for ECH, it seems clear to me that the LB would have to be the "client-facing server" in Fig. 2 of the ESNI draft. I don't know that this draft has to get into the details of how to implement arbitrary routing strategies. WDYT?

Reviewer 2: OK.

### Baseline Fallback Algorithm

All load balancers MUST implement a baseline fallback algorithm that takes only
the 4-tuple as an input and outputs a routing decision.

If it is impossible for the server to generate CIDs that the load balancer
cannot decode (see {{config-failover}}), there are no further requirements in
this subsection.

Otherwise, the load balancer SHOULD maintain a table of 4-tuples that carried
unroutable DCIDs and the resulting routing decision. Provided the table does
not overflow, and the load balancer does not lose state, this allows connections
to survive when the server pool changes, which would sometimes change the output
of the fallback algorithm.

The load balancer MAY maintain a table of observed unroutable DCIDs and the
resulting routing decision. Provided the table does not overflow, these
connections will be robust to NAT rebinding.

Load balancers SHOULD maintain per-flow timers to periodically purge state in
the tables described above.

### Advanced Fallback Algorithm

Some architectures may require a load balancer to choose a server pool based
on deep packet inspection of a client packet. For example, it may use the TLS

> 1.3 Server Name Indication (SNI) field. The advanced fallback algorithm enables

Reviewer: Should this be a link to https://www.rfc-editor.org/rfc/rfc6066#section-3 ?

an unroutable CID for the reasons described in {{config-failover}}.

Even if the fallback algorithm solely uses the 4-tuple, it MAY store the
algorithm result in case changes to the server pool change fallback algoritm
output.

The load balancer MAY also store the routing decision indexed by DCID. This
is not robust to new server-selected DCIDs, but can increase robustness when
a connection to a server generating unroutable CIDs experiences a NAT rebinding.
An entry can be freed when the corresponding 4-tuple uses a different CID,
whether routable or not.

### Load Balancer Forwarding {#load-balancer-forwarding}

Packets with routable CIDs are, of course, routed in accordance with the
decoded server ID.

For unroutable CIDs, load balancers execute the following steps in order
until one results in a routing decision.

1. If the load balancer has a table of routing decisions by DCID, and the
DCID is present in it, route the packet accordingly.
1. If the load balancer has a table of routing decisions by 4-tuple, and
the DCID is present in it, route the packet accordingly.
1. Use the fallback algorithm to make a routing decision and record the results

> in the tables indexed by 4-tuple and/or DCID.

Author: As stated a few paragraphs above, there is never a case where the table has no value. If the hash includes the DCID, the server could change the DCID.

Even if fully 4-tuple based, if the server pool changes, the a 4-tuple hash could give you a different result.

However, "if applicable" is a reasonable thing to add.

When load balancers and servers are guaranteed to have common non-null
configuration, all valid QUIC short headers have routable DCIDs. Under these
conditions, load balancers MAY drop packets with short headers and unroutable
DCIDs.

When load balancers are guaranteed to have a superset of all server
configurations, but servers might have no configuration at all, all valid QUIC
short headers have either a routable DCID or use config ID 0b111. Under these
conditions, load balancers MAY drop packets with short headers and unroutable
DCIDs that do not have config ID 0b111.

Load balancers MAY drop packets with long headers and unroutable DCIDs if
and only if it knows that the encoded QUIC version does not allow an unroutable
DCID in a packet with that signature. For example, a load balancer can safely
drop a QUIC version 1 Handshake packet with an unroutable DCID, as a
version 1 Handshake packet sent to a QUIC-LB routable server will always have
a server-generated routable CID. Load balancers MUST NOT drop packets with
long headers and unknown QUIC versions.

Under all other circumstances, load balancers MUST NOT drop packets with

> unroutable DCIDs.

Author: Welp, this language is broken now that 0b111 is Unroutable and server-generated.

> unroutable DCIDs.

Author: I am thinking about eliminating all the language about dropping packets. It seems to have a very narrow application and is easy to circumvent if an attacker is really trying to launch a DoS (just send a mass of Initial packets with different unroutable CIDs).

This would also simplify the version-invariance considerations further down. WDYT?

> unroutable DCIDs.

Author: I filed #238 to delete all that text.

QUIC connection handshake. Another might simply hash the address/port 4-tuple.
See also {{version-invariance}}.

Note that any information used by the Fallback Algorithm might not be repeated
from one QUIC packet to another. The 4-tuple is assumed to remain stable during
the handshake, but the server can change the destination connection ID after the
client's first flight.

> 

Reviewer 2: Sure. But if the server changes the CID, isn't it expected to pick a routable CID?

Author: If it doesn't have a config, it uses 0b111, which is now defined as unroutable. That's the big change in this PR.

Reviewer 2: OK.

Therefore, if the fallback algorithm uses information other than the 4-tuple in
its routing decision, the load balancer SHOULD store the routing decision for
that 4-tuple in order to route future packets unless doing so would exceed
resource limits. It can free these records when the 4-tuple uses a routable CID.
This state allows the load balancer to successfully route other packets from
the client's first flight, as well as subsequent flights if the server provides
an unroutable CID for the reasons described in {{config-failover}}.

> 

Reviewer 2: Do LBs need to worry about RSS in TLS? Or about stateless Retry?

Author: How so? IIRC those things won't change the 4-tuple.  By either using a pure 4-tuple hash, or a table of results by 4-tuple, the LB will get consistent results over handshake timescales, unless it uses a table and the table overflows.

Reviewer 2: Stateless retry will not change the 4 tuple, but it will change the DCID. The client that receives the retry will get a new DCID suggested by the server. The response will come from that new DCID. If the server is not configured yet, that new DCID will be "unroutable". This new unroutable DCID will be coming from the same 4-tuple, but the hashing of four tuples to servers may change between the time the IDCID was first seen and the new DCID is first seen.

RSS forces a kind of handshake between client and server. The client will switch from using the IDCID to using as DCID that the CID server used as a source CID. The same handshake will happen if the server sends an initial ACK before the client has sent all initial packets, after the client receives the first packets from the server. Same risk: if the hash table changes, the transition to the new DCID will fail.

Basically, if we route per hash of 4 tuple, the routing fails when hash changes. If the LB remembers the DCID, the routing fails if both the DCID and the hash table change. Remembering the DCID in the global table improves the robustness generally, with the possible exception of the IDCID (routing of initial packets). It might be better to not insert those in any table. In case of NAT rebinding, the connection only survives if the LB remembers the DCID in the global table.

Author: For stateless retry, we have to split this into the happy path, the unhappy path, and the very unhappy path, and the very very unhappy path.

In the happy path, the Retry packet provides a routable CID, and there is no problem.

In the unhappy path, there is no common configuration, so the server's CID is unroutable. The client's second attempt, it uses the same 4-tuple almost immediately, and the rest of the packet is pretty much the same. The second packet is a hit in the 4-tuple table, so it goes to the same place.

In the very unhappy path, there is no common config and the 4-tuple table is missing or overflowing (maybe reasonable in a Retry case). Pretty much everything in the second packet is the same except the DCID, so any fallback algorithm that does not use DCID should work. One that uses DCID will probably fail.

In the very very unhappy path, there is not common config, and the client delays the second attempt, long enough that there the hash result changes or there is a NAT rebinding, and/or the 4-tuple table entry times out. I would consider this edge case to be pathological behavior.

If by "RSS" you mean "HRR", I don't see that it's functionally different. The HRR will provide a new CID that has similar QUIC-LB properties.

Finally, this is all presumptive of v1/v2. Many other patterns are possible.

Even if the fallback algorithm solely uses the 4-tuple, it MAY store the
algorithm result in case changes to the server pool change fallback algorithm
output.

> 

Reviewer 2: I am not an LB implementer, but aren't you describing some kind of race condition? Once the server has received the first flight, it will provide its source CID to the client. The next packets from the client will use that as a DCID, and it will be routable.

Author: Again, all of this is covering the case when the server does not have a config and is using 0b111.

Reviewer 2: OK.

The load balancer MAY also store the routing decision indexed by DCID. This
is not robust to new server-selected DCIDs, but can increase robustness when
a connection to a server generating unroutable CIDs experiences a NAT rebinding.
An entry can be freed when the corresponding 4-tuple uses a different CID,
whether routable or not.

> 

Reviewer 2: We seem to be dancing around the issue. First, NAT rebinding *during the client's first fly" ought to be fairly rare -- rare enough to be treated as an error case. Second, resilience to NAT rebinding requires routing based on DCID, instead of the 4 tuple. Why not just say that if a server uses unroutable CIDs, then NAT rebinding is not supported? And maybe tell servers that use unroutable CIDs to set the TP "disable_active_migration=TRUE" ?

Author: Whoops, the language about disable_active_migration and NEW_CONNECTION_ID fell out of the edit. I've restored it.

### Load Balancer Forwarding {#load-balancer-forwarding}

Packets with routable CIDs are, of course, routed in accordance with the
decoded server ID.

For unroutable CIDs, load balancers execute the following steps in order
until one results in a routing decision.

1. If the load balancer has a table of routing decisions by DCID, and the
DCID is present in it, route the packet accordingly.
1. If the load balancer has a table of routing decisions by 4-tuple, and
the DCID is present in it, route the packet accordingly.
1. Use the fallback algorithm to make a routing decision and, if applicable, record the results
in the tables indexed by 4-tuple and/or DCID.

> 

Reviewer 2: Yes. Note that it is much easier to discuss Fallback Algorithms *after* we have defined the default behavior. Consider changing the order of these sections.

# Acknowledgments

Manasi Deval, Erik Fuller, Toma Gavrichenkov, Greg Greenway, Jana Iyengar,
Subodh Iyengar, Stefan Kolbl, Ladislav Lhotka, Jan Lindblad, Ling Tao Nju,
Ilari Liusvaara, Kazuho Oku, Udip Pant, Zaheduzzaman Sarker, Ian Swett, Andy
Sykes, Martin Thomson, Dmitri Tikhonov, Victor Vasiliev, Xingcan Lan, Yu Zhu,
and William Zeng Ke all provided useful input to this document.

# Change Log

> **RFC Editor's Note:**  Please remove this section prior to
> publication of a final version of this document.

## since draft-ietf-quic-load-balancers-20

- Changed defintion of Unroutable DCIDs, and rewrote sections on config
failover and fallback routing to avoid misrouted connections.

> 

Reviewer 2: Typo, defintion -> definition
